torch/csrc/autograd/variable.cpp

```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
   return get_autograd_meta(self);
 }
 
+void update_retains_grad(
+    const at::TensorBase& self,
+    const std::shared_ptr<torch::autograd::Node>& fn) {
+  if (!self.retains_grad()) {
+    return;
+  }
+  auto idx = impl::get_autograd_meta(self)->retains_grad_;
+  auto old_list = materialize_autograd_meta(self)->cpp_hooks_list_;
+  auto retains_grad_hook = (*old_list)[idx];
+  auto new_list = std::make_shared<hooks_list>();
+  new_list->push_back(retains_grad_hook);
+  (*old_list)[idx] = nullptr;
+  materialize_autograd_meta(self)->cpp_hooks_list_ = new_list;
+
+  std::unique_ptr<FunctionPreHook> hook_ptr(
```

Comment:
Why not reset the `AM.retains_grad_` field and call `self.retain_grad()` here? That will avoid duplication.

Reply:
retain_grad would call into register_hook which would wrap the function in a lambda again